### PR TITLE
feat(llm): support multi reranker & enhance the UI

### PR DESCRIPTION
**Pull request by jasinliu**
As title.

### CODE REVIEW

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 1]
```diff
@@ -289,48 +354,76 @@ def init_rag_ui() -> gr.Interface:
...
+                        gr.Textbox(
+                            value=settings.openai_api_key,
+                            label="api_key",
+                            type="password",
+                        ),
                         gr.Textbox(value=settings.openai_api_base, label="api_base"),
-                        gr.Textbox(value=settings.openai_language_model, label="model_name"),
+                        gr.Textbox(
+                            value=settings.openai_language_model, label="model_name"
+                        ),
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:382]: Do we need expand **each function** param to multi lines?

File: hugegraph-llm/src/hugegraph_llm/api/rag_api.py [hunk 2]
```diff
@@ -45,22 +63,50 @@ def llm_config_api(req: LLMConfigRequest):
...
+                req.host, req.port, req.language_model, origin_call="http"
+            )
+        return generate_response(RAGResponse(status_code=res, message="Missing Value"))
+
+    @router.post("/config/rerank", status_code=status.HTTP_201_CREATED)
+    def rerank_config_api(req: RerankerConfigRequest):
+        settings.reranker_type = req.reranker_type
+
+        if req.reranker_type == "cohere":
+            res = apply_reranker_conf(
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/api/rag_api.py:109]: `res` might be `Null` here? Maybe define it in advance?

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 3]
```diff
@@ -72,7 +82,9 @@ def rag_answer(
     if graph_search:
         searcher.extract_keyword().match_keyword_to_id().query_graph_for_rag()
     # TODO: add more user-defined search strategies
-    searcher.merge_dedup_rerank().synthesize_answer(
+    searcher.merge_dedup_rerank(
+        graph_ratio, rerank_method, near_neighbor_first, custom_related_information
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:86]: `rerank_method` is `str`, but expected type  is `Literal["bleu", "reranker"]`

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 4]
```diff
@@ -228,6 +246,37 @@ def apply_embedding_config(arg1, arg2, arg3, origin_call=None) -> int:
     return status_code
 
 
+def apply_reranker_config(arg1, arg2, arg3: str | None = None, origin_call=None) -> int:
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:249]: if `arg1~n` is not flexible, better to use a real name for it (like `api_key`...)

File: hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py [hunk 5]
```diff
@@ -16,59 +16,119 @@
...
+                reranker.get_rerank_lists(query, vertex_degree) + [""] for vertex_degree in vertex_degree_list
+            ]
+        depth = len(vertex_degree_list)
+        for result in results:
+            if result not in knowledge_with_degree:
+                knowledge_with_degree[result] = [result] + [""] * (depth - 1)
+            if len(knowledge_with_degree[result]) < depth:
+                knowledge_with_degree[result] += [""] * (depth - len(knowledge_with_degree[result]))
+
+        def sort_key(result: str) -> Tuple[int, ...]:
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py:130]: `result` is used in global, better to change the name

File: hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py [hunk 6]
```diff
@@ -16,59 +16,119 @@
...
+        assert method in [
+            "bleu",
+            "reranker",
+        ], f"Unimplemented rerank method '{method}'."
         self.embedding = embedding
+        self.graph_ratio = graph_ratio
         self.topk = topk
-        if strategy == "bleu":
-            self.rerank_func = self._bleu_rerank
-        elif strategy == "priority":
```
[the remark this answers is not in the document]
Comment (jasinliu) [reply hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py:43]: > Maybe we need to keep this strategy and implement it in the future?

I can reserve a parameter `priority: bool` for it. Does this work?

[the remark this answers is not in the document]
Comment (simon824) [reply hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py:43]: Do you mean that this priority parameter applies to all reranking methods?  If so, it makes sense.

[the remark this answers is not in the document]
Comment (jasinliu) [reply hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py:43]: > Do you mean that this priority parameter applies to all reranking methods? If so, it makes sense.

Is this ok?

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 7]
```diff
@@ -229,6 +246,42 @@ def apply_embedding_config(arg1, arg2, arg3, origin_call=None) -> int:
     return status_code
 
 
+def apply_reranker_config(
+    reranker_api_key: str | None = None,
```
Comment (ChenZiHong-Gavin) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:250]: get error on py3.9
image

Comment (jasinliu) [reply hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:250]: > get error on py3.9 image

Does `Optional` fix it?

File: hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py [hunk 8]
```diff
@@ -16,59 +16,134 @@
...
+        if self.method == "reranker":
+            reranker = Rerankers().get_reranker()
+            return reranker.get_rerank_lists(query, results, topn)
+
+    def _rerank_with_vertex_degree(
+        self,
+        query: str,
+        results: List[str],
+        topn: int,
+        vertex_degree_list: List[List[str]] | None,
```
Comment (ChenZiHong-Gavin) [inline hugegraph-llm/src/hugegraph_llm/operators/common_op/merge_dedup_rerank.py:113]: also here, use `Union` instead?

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py (hunk 7 above)
Comment (ChenZiHong-Gavin) [reply hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:250]: yes, `Optional` would be fine

File: hugegraph-llm/src/hugegraph_llm/operators/graph_rag_task.py [hunk 9]
```diff
@@ -73,26 +75,26 @@ def extract_keyword(
...
     def match_keyword_to_id(
-            self,
-            by: Literal["query", "keywords"] = "keywords",
-            topk_per_keyword: int = 1,
-            topk_per_query: int = 10
+        self,
+        by: Literal["query", "keywords"] = "keywords",
+        topk_per_keyword: int = 1,
+        topk_per_query: int = 10,
     ):
```
[the remark this answers is not in the document]
Comment (jasinliu) [reply hugegraph-llm/src/hugegraph_llm/operators/graph_rag_task.py:107]: > indent twice here

The indent length is now sufficient

File: hugegraph-llm/src/hugegraph_llm/operators/hugegraph_op/graph_rag_query.py
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/operators/hugegraph_op/graph_rag_query.py:1]: TODO: we need to simplify the process logic soon

**Review by imbajin — APPROVED** [review]

**Pull request merged**